app/assets/javascripts/discourse/app/controllers/reorder-categories.js

```diff
@@ -63,56 +63,102 @@ export default Controller.extend(ModalFunctionality, Evented, {
+      }
     } else {
-      // Find category occupying target position
-      otherCategory = this.categoriesOrdered.find(
-        (c) => c.get("position") === category.get("position") + direction
+      // Moving up (position gets smaller)
+      const otherCategory = this.categoriesOrdered.find(
+        (c) =>
+          // find category currently at targetPosition
+          c.get("position") === targetPosition
       );
+      if (otherCategory && otherCategory.get("ancestors")) {
+        // Target category is a subcategory, adjust targetPosition to account for ancestors
+        const highestAncestor = otherCategory
+          .get("ancestors")
+          .reduce((current, min) =>
+            current.get("position") < min.get("position") ? current : min
+          );
+        targetPosition = highestAncestor.get("position");
+      }
     }
 
-    if (otherCategory) {
-      // Try to swap positions of the two categories
-      if (category.get("id") !== otherCategory.get("id")) {
-        const currentPosition = category.get("position");
-        category.set("position", otherCategory.get("position"));
-        otherCategory.set("position", currentPosition);
-      }
-    } else if (direction < 0) {
-      category.set("position", -1);
-    } else if (direction > 0) {
-      category.set("position", this.categoriesOrdered.length);
+    // Adjust target position for range bounds
+    if (targetPosition >= this.categoriesOrdered.length) {
+      // Set to max
+      targetPosition = this.categoriesOrdered.length - 1;
+    } else if (targetPosition < 0) {
+      // Set to min
+      targetPosition = 0;
     }
 
+    // Update other categories between current and target position
+    this.categoriesOrdered.map((c) => {
+      if (direction < 0) {
+        // Moving up (position gets smaller)
+        if (
+          c.get("position") < category.get("position") &&
+          c.get("position") >= targetPosition
+        ) {
+          const newPosition = c.get("position") + 1;
+          c.set("position", newPosition);
+        }
+      } else {
+        // Moving down (position gets larger)
+        if (
+          c.get("position") > category.get("position") &&
+          c.get("position") <= targetPosition
+        ) {
+          const newPosition = c.get("position") - 1;
+          c.set("position", newPosition);
+        }
+      }
+    });
+
+    // Update this category's position to target position
+    category.set("position", targetPosition);
+
     this.reorder();
   },
 
   actions: {
     change(category, event) {
-      let newPosition = parseInt(event.target.value, 10);
-      newPosition = Math.min(
-        Math.max(newPosition, 0),
-        this.categoriesOrdered.length - 1
-      );
-
-      this.move(category, newPosition - category.get("position"));
+      let newPosition = parseFloat(event.target.value);
+      let direction = newPosition - category.get("position");
+      if (direction < 0) {
```

Comment:
There is no need to compute direction twice just to adjust the position.

```suggestion
      if (newPosition < category.get("position")) {
```

Eventually, you could write the whole block such as:

```javascript
newPosition = newPosition < category.get("position") ? Math.ceil(newPosition) : Math.floor(newPosition);
```

Reply:
This is much better. Thanks!